### PR TITLE
PUB-507 Alphabetical Search

### PR DESCRIPTION
**Pull request by ChrisS1512**
### JIRA link ###

https://tools.hmcts.net/jira/browse/PUB-507

### Change description ###

This change adds the alphabetical view, which lets users search for courts, and click through to the hearing list page.

The following are the key bits of functionality:

- Find a court or Tribunal listing (with alphabetical search bar to select name of court or tribunal running along the top)
- There is an A-Z list with the courts/tribunals alongside and on the RHS number of hearings in that court/tribunal on that day
- When user clicks on a letter, they are taken to the courts and tribunals starting with that letter
- If the user clicks on the 'back to the top' link, then they are taken back to the top of the page where the alphabetical bar is displayed
- If there are no hearings on that day for a certain letter then show 'No hearings are scheduled in any of these locations today' which is greyed out and not clickable

Note - Getting to this page needs to be done directly at https://localhost:8080/alphabetical-search

An screenshot example with test data is shown below:

Screenshot 2021-08-09 at 11 27 50

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/controllers/AlphabeticalSearchController.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    for (let i = 0; i < 26; i++) {
+      const letter = String.fromCharCode(65 + i);
+      alphabetArray[letter] = {};
+    }
+
+    courtsList = new InputFilterService().alphabetiseResults(courtsList, 'name');
+
+    //Then loop through each court, and add it to the list
+    courtsList.forEach(item => {
+      if (item.hearings != 0) {
```
[the remark this answers is not in the document]
Comment (ChrisS1512) [reply src/main/controllers/AlphabeticalSearchController.ts]: Good spot. I've updated this now to '!=='

**Review by commitBlob — APPROVED** [review]

File: src/main/views/alphabetical-search.njk [hunk 2]
```diff
@@ -0,0 +1,81 @@
+{% from "govuk/components/table/macro.njk" import govukTable as 'govukTable' %}
+
+{% extends "template.njk" %}
+{% block pageTitle %}
+  {{ "Court List" | safe }}
+{% endblock %}
+{% block beforeContent %}
+  {% from "govuk/components/back-link/macro.njk" import govukBackLink %}
```
Comment (kurt-uk) [inline src/main/views/alphabetical-search.njk:8]: we have the common components macro that we import for back buttons and/or submit buttons on generic pages
use:
```
{{ goBack('/search-option') }}
```

**Review by kurt-uk — CHANGES_REQUESTED** [review]
dont forget the e2e

**Comment by kurt-uk** [comment]
youve forgotten to link the page from the search option page

**Comment by ChrisS1512** [comment]
> youve forgotten to link the page from the search option page

I'll need to wait for the search-option PR (506) to go in for this one - as this branch doesn't have the form submission functionality in there

File: src/main/views/alphabetical-search.njk (hunk 2 above)
Comment (commitBlob) [reply src/main/views/alphabetical-search.njk:8]: Agree with this although 507 will fail without 506 in that case as common components were defined on the 506 branch

Comment (ChrisS1512) [reply src/main/views/alphabetical-search.njk]: Merged in 506 - and now updated to use the common components back button

**Comment by ChrisS1512** [comment]
> > youve forgotten to link the page from the search option page
> 
> I'll need to wait for the search-option PR (506) to go in for this one - as this branch doesn't have the form submission functionality in there

I've now updated this to link from the search-option page, to the alphabetical search page

File: src/main/controllers/SearchOptionController.ts [hunk 3]
```diff
@@ -9,5 +9,8 @@ export default class SearchOptionController {
     if (req.body['find-choice'] === 'search') {
       res.redirect('search');
     }
+    if (req.body['find-choice'] === 'find') {
```
Comment (kurt-uk) [inline src/main/controllers/SearchOptionController.ts:12]: theres also the unit test for this path too in SearchOptionController.test.ts

Comment (ChrisS1512) [reply src/main/controllers/SearchOptionController.ts:12]: Added unit test in for this

File: src/main/resources/actions/hearingActions.ts [hunk 4]
```diff
@@ -3,20 +3,18 @@ import path from 'path';
 
 export class HearingActions {
   mocksPath = '../mocks/';
-  
+
```
Comment (kurt-uk) [inline src/main/resources/actions/hearingActions.ts:6]: 🙏🏼

File: src/main/views/alphabetical-search.njk [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+{% from "./macros/common-components.njk" import goBack %}
+
+{% extends "template.njk" %}
+{% block pageTitle %}
+  {{ "Court List" | safe }}
+{% endblock %}
+{% block beforeContent %}
+  {{ goBack('/search-option') }}
+{% endblock %}
+{% block content %}
```
Comment (kurt-uk) [inline src/main/views/alphabetical-search.njk:11]: this section is so elegant

**Review by kurt-uk — APPROVED** [review]
👍🏼

**Pull request merged**